django/db/models/fields/related.py

```diff
@@ -731,26 +732,23 @@ def get_reverse_path_info(self):
         pathinfos = [PathInfo(from_opts, opts, (opts.pk,), self.remote_field, not self.unique, False)]
         return pathinfos
 
-    def get_lookup(self, lookup_name):
-        if lookup_name == 'in':
-            return RelatedIn
-        elif lookup_name == 'exact':
-            return RelatedExact
-        elif lookup_name == 'gt':
-            return RelatedGreaterThan
-        elif lookup_name == 'gte':
-            return RelatedGreaterThanOrEqual
-        elif lookup_name == 'lt':
-            return RelatedLessThan
-        elif lookup_name == 'lte':
-            return RelatedLessThanOrEqual
-        elif lookup_name == 'isnull':
-            return RelatedIsNull
-        else:
-            raise TypeError('Related Field got invalid lookup: %s' % lookup_name)
-
-    def get_transform(self, *args, **kwargs):
-        raise NotImplementedError('Relational fields do not support transforms.')
+    def _get_lookup(self, lookup_name):
+        try:
+            return self.class_lookups[lookup_name]
+        except KeyError:
+            # To allow for inheritance, check parent class' class_lookups.
+            for parent in inspect.getmro(self.__class__):
+                # Related lookups need to be explicitly registered to related fields.
+                if parent is ForeignObject:
+                    break
+                if 'class_lookups' not in parent.__dict__:
```

Comment:
Thanks for running some timing on these. We probably don't focus on performance as much as we should.

That's a fairly significant increase in time for the non related field (nearly 3 times slower). 50% slower for related field is not great either. I'm not sure that caching the bases for each field type is something we can do without busting the cache in register_lookup. I haven't put enough thought into how this might work to offer a great solution.

What I was thinking of was something like:

```
def get_lookups(self):
        cached = getattr(self, 'cached_lookups')
        if cached is not None: return cached
        class_lookups = [getattr(cls, 'class_lookups', {}) for cls in inspect.getmro(self.__class__)]
        merged = merge_dicts(class_lookups)
        self.cached_lookups = merged
        return merged
```

Would have to be careful that this code cached on the correct class though, which is probably unlikely considering the placement of RegisterLookupMixin to the left.

What I was hoping to avoid was recursively searching the base hierarchy for every access of a lookup.

```Model.objects.filter(intfield__gt=5, intfield__lt=10)``` would search twice. It'd be a lot nicer if the second lookup (`lt`) was an index into a saved dict rather than another base search.

If there's no obvious way to add caching then I think it best to go with the quicker solution (assuming the quicker was your option 1 and not just the status quo.

@carljm I know you're working with performance data lately - have you got any comments here? Do you have any idea if this kind of overhead would be noticeable in big deployments?